File: sdk/samples/01_basics/cpp/transport_select.cpp
// SPDX-License-Identifier: Apache-2.0 OR MIT
// Copyright (c) 2025-2026 naskel.com

/**
 * HDDS Sample: Transport Selection (C++)
 *
 * Demonstrates creating participants with explicit transport selection.
 * Shows UdpMulticast (default) and IntraProcess modes.
 *
 * Build:
 *     cd build && cmake .. && make transport_select
 *
 * Usage:
 *     ./transport_select              # Default UDP multicast transport
 *     ./transport_select intra        # IntraProcess transport
 *     ./transport_select udp          # Explicit UDP multicast transport
 *
 * Expected output:
 *     [OK] Participant created with udp transport
 *     [SENT] Transport test message #1
 *     ...
 *
 * Key concepts:
 * - Default transport is UDP multicast (LAN discovery)
 * - IntraProcess transport for same-process communication (zero-copy)
 * - Transport selected at participant creation
 */

#include <hdds.hpp>
#include <iostream>
#include <string>
#include <thread>
#include <chrono>
#include <cstring>

#include "generated/HelloWorld.hpp"

using namespace hdds_samples;
using namespace std::chrono_literals;

constexpr int NUM_MESSAGES = 5;

int main(int argc, char **argv) {
    std::string transport = "udp";
    if (argc > 1) {
        transport = argv[1];
    }

    std::cout << std::string(60, '=') << "\n";
    std::cout << "Transport Selection Demo\n";
    std::cout << "Selected transport: " << transport << "\n";
    std::cout << std::string(60, '=') << "\n\n";

    try {
        hdds::logging::init(hdds::LogLevel::Warn);

        std::cout << "--- Available Transports ---\n";
        std::cout << "  udp    - UDP multicast (default, LAN discovery)\n";
        std::cout << "  intra  - IntraProcess (same-process, zero-copy)\n";
        std::cout << "\n";

        // Create participant with selected transport
        hdds::TransportMode mode = (transport == "intra")
            ? hdds::TransportMode::IntraProcess
            : hdds::TransportMode::UdpMulticast;

        hdds::Participant participant("TransportDemo", mode);

        std::cout << "[OK] Participant created with " << transport << " transport\n";

        // Create endpoints
        auto writer = participant.create_writer_raw("TransportTopic");
        std::cout << "[OK] DataWriter created on 'TransportTopic'\n";

        auto reader = participant.create_reader_raw("TransportTopic");
        std::cout << "[OK] DataReader created on 'TransportTopic'\n\n";

        // Send messages
        std::cout << "--- Sending " << NUM_MESSAGES << " messages via "
                  << transport << " ---\n\n";

        for (int i = 0; i < NUM_MESSAGES; i++) {
            HelloWorld msg(i + 1, "Transport test #" + std::to_string(i + 1) +
                                   " (" + transport + ")");
            std::uint8_t buffer[4096];
            int bytes = msg.encode_cdr2_le(buffer, sizeof(buffer));
            if (bytes > 0) {
                writer->write_raw(buffer, static_cast<size_t>(bytes));
            }
            std::cout << "[SENT] id=" << msg.id << " msg='" << msg.message << "'\n";
            std::this_thread::sleep_for(200ms);
        }

        // Read back
        std::cout << "\n--- Reading messages ---\n\n";

        hdds::WaitSet waitset;
        waitset.attach(reader->get_status_condition());

        if (waitset.wait(2s)) {
            while (auto data = reader->take_raw()) {
                HelloWorld msg;
                if (msg.decode_cdr2_le(data->data(), data->size()) > 0) {
                    std::cout << "[RECV] id=" << msg.id
                              << " msg='" << msg.message << "'\n";
                }
            }
        } else {
            std::cout << "[TIMEOUT] No messages received "
                      << "(run two instances to test)\n";
        }

    } catch (const hdds::Error& e) {
        std::cerr << "HDDS Error: " << e.what() << std::endl;
        return 1;
    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
        return 1;
    }

    std::cout << "\n=== Transport Selection Complete ===\n";
    return 0;
}

File: sdk/samples/01_basics/python/transport_select.py
#!/usr/bin/env python3
# SPDX-License-Identifier: Apache-2.0 OR MIT
# Copyright (c) 2025-2026 naskel.com

"""
HDDS Sample: Transport Selection (Python)

Demonstrates creating participants with explicit transport selection.
Shows UDP (default), TCP, and how to switch between transports.

Usage:
    python transport_select.py              # Default UDP transport
    python transport_select.py tcp          # TCP transport
    python transport_select.py udp          # Explicit UDP transport

Expected output:
    [OK] Participant created with UDP transport
    [SENT] Transport test message #1
    ...

Key concepts:
- Default transport is UDP multicast
- TCP transport for NAT traversal / WAN
- Transport selected at participant creation
"""

import os
import sys
import time

sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', '..', '..', 'python'))

import hdds

NUM_MESSAGES: int = 5


def main() -> int:
    transport = "udp"
    if len(sys.argv) > 1:
        transport = sys.argv[1].lower()

    print("=" * 60)
    print("Transport Selection Demo")
    print(f"Selected transport: {transport}")
    print("=" * 60)
    print()

    hdds.logging.init(hdds.LogLevel.INFO)

    print("--- Available Transports ---")
    print("  udp  - UDP multicast (default, LAN discovery)")
    print("  tcp  - TCP point-to-point (NAT traversal, WAN)")
    print()

    # Create participant with selected transport
    if transport == "tcp":
        kind = hdds.TransportKind.TCP
    else:
        kind = hdds.TransportKind.UDP

    participant = hdds.Participant("TransportDemo", transport=kind)
    print(f"[OK] Participant created with {transport.upper()} transport")

    # Create endpoints
    writer = participant.create_writer("TransportTopic")
    print("[OK] DataWriter created on 'TransportTopic'")

    reader = participant.create_reader("TransportTopic")
    print("[OK] DataReader created on 'TransportTopic'\n")

    # Send messages
    print(f"--- Sending {NUM_MESSAGES} messages via {transport} ---\n")

    for i in range(NUM_MESSAGES):
        msg_id = i + 1
        payload = f"Transport test #{msg_id} ({transport})".encode('utf-8')
        writer.write(payload)
        print(f"[SENT] id={msg_id} msg='Transport test #{msg_id} ({transport})'")
        time.sleep(0.2)

    # Read back
    print("\n--- Reading messages ---\n")

    waitset = hdds.WaitSet()
    waitset.attach_reader(reader)

    if waitset.wait(timeout=2.0):
        while True:
            data = reader.take()
            if data is None:
                break
            print(f"[RECV] {data.decode('utf-8')}")
    else:
        print("[TIMEOUT] No messages received (run two instances to test)")

    participant.close()

    print("\n=== Transport Selection Complete ===")
    return 0


if __name__ == "__main__":
    sys.exit(main())
